File: src/main/java/com/ycc/haechan/Solution1.java
package com.ycc.haechan;


public class Solution1 {

    public int solve(String s){
        int answer = s.length();

        for (int i = 1; i <= s.length() / 2; i++) {

            StringBuilder result = new StringBuilder();
            String presentValue = s.substring(0, i);
            int count = 1;

            for (int j = i; j < s.length(); j += i) {
                int len = Math.min(j + i, s.length());
                String substring = s.substring(j, len);
                if (presentValue.equals(substring)) {
                    count++;
                } else {
                    if (count == 1) {
                        result.append(presentValue);
                    } else {
                        result.append(count).append(presentValue);
                    }
                    presentValue = substring;
                    count = 1;
                }
            }

            if (count != 1) {
                result.append(count).append(presentValue);
            } else {
                result.append(presentValue);
            }

            answer = Math.min(result.length(), answer);

        }

        return answer;
    }
}

File: src/main/java/com/ycc/beomseok/Solution1.kt
package com.ycc.beomseok

class Solution1 {
    fun solve(compressed: String): Int {
        val compressedResult = mutableListOf<Int>()
        val maxWrapWordCount = compressed.length / 2
        for (i in maxWrapWordCount downTo 1) {
            val compress = compress(compressed, i)
            compressedResult += compress
        }
        return compressedResult.min()?: compressed.length
    }

    private fun compress(src: String, wrapWordCount: Int): Int {
        val wordArray = mutableListOf<String>()
        val wordCounterArray = mutableListOf<Int>()
        for (i in src.indices step wrapWordCount) {
            val endIndex = (i + wrapWordCount).coerceAtMost(src.length)
            val wrapWord = src.substring(i, endIndex)
            if (wordArray.isNotEmpty() && wordArray[wordArray.size - 1] == wrapWord) {
                wordCounterArray[wordArray.size - 1]++
            } else {
                wordArray.add(wrapWord)
                wordCounterArray.add(1)
            }
        }
        var result = 0
        for ((index, word) in wordArray.withIndex()) {
            val entryCount = if (wordCounterArray[index] == 1) "" else "${wordCounterArray[index]}"
            val wordCount = "$entryCount${word}".length
            result += wordCount
        }
        return result
    }
}